[mlir] Compose expand of collapse to cast

amrami opened the pull request:
In some cases `y = expand(collapse(x)` cannot be folded into x, since x and y have different types.
In that case, we check if the two types are cast compatible.
If they are, it means the two types have compatible shape and layout and y can be folded into cast(x).

This causes a change in memref::CastOp::areCastCompatible, where now a dim of size 1 may have different strides.

Groverkss reviewed (changes requested):
I don't think cast carries the same dimensional information as expand/collapse does. You cannot track the dimensions anymore.

MaheshRavishankar commented on mlir/test/Dialect/MemRef/canonicalize.mlir:
```diff
@@ -1193,6 +1193,25 @@ func.func @collapse_expand_fold_to_cast(%m: memref<?xf32, strided<[1]>, 3>, %sz0
 
+// CHECK-LABEL: func @expand_collapse_fold_to_cast(
+//  CHECK-SAME:     %[[m:.*]]: memref<1x1x384x384xui8, strided<[1179648, 147456, 384, 1]>>
+//       CHECK:   %[[casted:.*]] = memref.cast %[[m]] : memref<1x1x384x384xui8, strided<[1179648, 147456, 384, 1]>> to memref<1x1x384x384xui8, strided<[1179648, 1179648, 384, 1]>>
```
I am confused why there is even a cast here? This cast here seems like a nop-op.

MaheshRavishankar reviewed (changes requested):
The lit test confuses me. I expect that to just get folded away by the `ComposeExpandOfCollapseOp` pattern. Why is there a cast?

Commit:
CR

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -748,14 +748,18 @@ bool CastOp::areCastCompatible(TypeRange inputs, TypeRange outputs) {
-      for (const auto &aStride : enumerate(aStrides))
-        if (!checkCompatible(aStride.value(), bStrides[aStride.index()]))
+      for (const auto &[index, aStride] : enumerate(aStrides)) {
+        if (aT.getDimSize(index) == 1)
```
Is this valid for all dimensions, or just the leading dimensions?

Commit:
CR 2 - Adding a lit test where the mismatch in strides is in an internal dim

amrami commented:
ping

MaheshRavishankar reviewed (approved):
I think this makes sense. I am OK with this landing.

Commit:
[mlir] Compose expand of collapse to cast

In some cases expand(collapse(x) pair cannot be folded
into x, since it has different type than x.
In that case, it will be folded into cast.

This causes a change in memref::CastOp::areCastCompatible,
where now a dim of size 1 may have different strides.
